sql/core/src/main/scala/org/apache/spark/sql/dynamicpruning/PartitionPruning.scala

```diff
@@ -0,0 +1,264 @@
+    val overhead = otherPlan.collectLeaves().map(_.stats.sizeInBytes).sum.toFloat
+
+    filterRatio * partPlan.stats.sizeInBytes.toFloat > overhead.toFloat
+  }
+
+  /**
+   * Returns whether an expression is likely to be selective
+   */
+  private def isLikelySelective(e: Expression): Boolean = e match {
+    case Not(expr) => isLikelySelective(expr)
+    case And(l, r) => isLikelySelective(l) || isLikelySelective(r)
+    case Or(l, r) => isLikelySelective(l) && isLikelySelective(r)
+    case Like(_, _) => true
+    case _: BinaryComparison => true
+    case _: In | _: InSet => true
+    case _: StringPredicate => true
+    case _ => false
+  }
+
+  /**
+   * Search a filtering predicate in a given logical plan
+   */
+  private def hasSelectivePredicate(plan: LogicalPlan): Boolean = {
+    plan.find {
+      case f: Filter => isLikelySelective(f.condition)
+      case _ => false
+    }.isDefined
+  }
+
+  /**
+   * To be able to prune partitions on a join key, the filtering side needs to
+   * meet the following requirements:
+   *   (1) it can not be a stream
+   *   (2) it needs to contain a selective predicate used for filtering
+   *   (3) the join key is in the attribute set of the filtering plan
+   */
+  private def hasPartitionPruningFilter(key: Expression, plan: LogicalPlan): Boolean = {
+    !plan.isStreaming && hasSelectivePredicate(plan) &&
+      key.references.subsetOf(plan.outputSet)
+  }
+
+  private def canPruneLeft(joinType: JoinType): Boolean = joinType match {
+    case Inner | LeftSemi | RightOuter => true
+    case _ => false
+  }
+
+  private def canPruneRight(joinType: JoinType): Boolean = joinType match {
+    case Inner | LeftOuter => true
+    case _ => false
+  }
+
+  private def prune(plan: LogicalPlan): LogicalPlan = {
+    plan transformUp {
+      // skip this rule if there's already a DPP subquery on the LHS of a join
+      case j @ Join(Filter(_: DynamicPruningSubquery, _), _, _, _, _) => j
+      case j @ Join(_, Filter(_: DynamicPruningSubquery, _), _, _, _) => j
+      case j @ Join(left, right, joinType, Some(condition), hint) =>
+        var newLeft = left
+        var newRight = right
+
+        // extract the left and right keys of the join condition
+        val (leftKeys, rightKeys) = j match {
+          case ExtractEquiJoinKeys(_, lkeys, rkeys, _, _, _, _) => (lkeys, rkeys)
+          case _ => (Nil, Nil)
+        }
+
+        // checks if two expressions are on opposite sides of the join
+        def fromDifferentSides(x: Expression, y: Expression): Boolean = {
+          def fromLeftRight(x: Expression, y: Expression) =
+            !x.references.isEmpty && x.references.subsetOf(left.outputSet) &&
+              !y.references.isEmpty && y.references.subsetOf(right.outputSet)
+          fromLeftRight(x, y) || fromLeftRight(y, x)
+        }
+
+        splitConjunctivePredicates(condition).foreach {
+          case EqualTo(a: Expression, b: Expression)
```

Comment:
what about `EqualNullSafe`?

Reply:
Created a follow-up JIRA as https://issues.apache.org/jira/browse/SPARK-28959.

Reply:
Just think about it. Is EqualNullSafe different to EqualTo here? Can partition column be a null value? If it can not be, when building key is null, EqualTo's value is null, EqualNullSafe's value is false. For a Filter predicate, null is considered a false, isn't?

Reply:
partition column can be null, we will use a special string to represent null in the file path.

Reply:
oh, I see. `__HIVE_DEFAULT_PARTITION__`